src/platform/Darwin/DnssdContexts.cpp

```diff
@@ -0,0 +1,426 @@
+{
+    type     = ContextType::Resolve;
+    context  = cbContext;
+    callback = cb;
+    protocol = GetProtocol(cbAddressType);
+}
+
+ResolveContext::~ResolveContext()
+{
+    RemoveInterfaces();
+}
+
+void ResolveContext::DispatchFailure(DNSServiceErrorType err)
+{
+    ChipLogError(DeviceLayer, "Resolve (%s)", Error::ToString(err));
+    callback(context, nullptr, Span<Inet::IPAddress>(), CHIP_ERROR_INTERNAL);
+    MdnsContexts::GetInstance().Remove(this);
+}
+
+void ResolveContext::DispatchSuccess()
+{
+    for (auto & service : services)
+    {
+        auto & ips = addresses[service.first];
+
+        // Some interface may not have any ips, just ignore them.
+        if (ips.size() == 0)
+        {
+            continue;
+        }
+
+        // Use the first IP we got for the DnssdService.
+        service.second.mAddress.SetValue(ips.front());
+        ips.erase(ips.begin());
+        callback(context, &service.second, Span<Inet::IPAddress>(ips.data(), ips.size()), CHIP_NO_ERROR);
+        MdnsContexts::GetInstance().Remove(this);
+        return;
+    }
+}
+
+CHIP_ERROR ResolveContext::OnNewAddress(uint32_t interfaceId, const struct sockaddr * address)
+{
+    chip::Inet::IPAddress ip;
+    ReturnErrorOnFailure(chip::Inet::IPAddress::GetIPAddressFromSockAddr(*address, ip));
+    addresses[interfaceId].push_back(ip);
+
+#ifdef CHIP_DETAIL_LOGGING
+    char addrStr[INET6_ADDRSTRLEN];
+    ip.ToString(addrStr, sizeof(addrStr));
+    ChipLogDetail(DeviceLayer, "Mdns: %s interface: %" PRIu32 " ip:%s", __func__, interfaceId, addrStr);
+#endif // CHIP_DETAIL_LOGGING
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ResolveContext::OnNewLocalOnlyAddress()
+{
+    sockaddr_in6 sockaddr;
+    memset(&sockaddr, 0, sizeof(sockaddr));
+    sockaddr.sin6_len    = sizeof(sockaddr);
+    sockaddr.sin6_family = AF_INET6;
+    sockaddr.sin6_addr   = in6addr_loopback;
+    sockaddr.sin6_port   = htons((unsigned short) services[kDNSServiceInterfaceIndexLocalOnly].mPort);
+
+    return OnNewAddress(kDNSServiceInterfaceIndexLocalOnly, reinterpret_cast<struct sockaddr *>(&sockaddr));
+}
+
+bool ResolveContext::HasAddress()
+{
+    for (auto & interfaceAddresses : addresses)
+    {
+        if (interfaceAddresses.second.size())
+        {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+void ResolveContext::OnNewInterface(uint32_t interfaceId, const char * fullname, const char * hostnameWithDomain, uint16_t port,
+                                    uint16_t txtLen, const unsigned char * txtRecord)
+{
+    ChipLogDetail(DeviceLayer, "Mdns : %s hostname:%s fullname:%s interface: %" PRIu32, __func__, hostnameWithDomain, fullname,
+                  interfaceId);
+
+    DnssdService service = {};
+    service.mPort        = ntohs(port);
+    service.mInterface   = Inet::InterfaceId(interfaceId);
+
+    auto hostname = GetHostNameWithoutDomain(hostnameWithDomain);
```

Comment:
Please document why we are doing this?  The old code didn't seem to strip off the domain, afaict...

Reply:
The old code was wrong basically. If you look at https://github.com/project-chip/connectedhomeip/blob/master/src/lib/dnssd/platform/Dnssd.h#L67 you can see that the size of the field is `kHostNameMaxLength + 1` where `kHostNameMaxLength` is https://github.com/project-chip/connectedhomeip/blob/c2b6479b7aa27517f36b97173efc3e318f0937a9/src/lib/dnssd/Constants.h#L32

So the field is sized to contains a 12 bytes Mac address or an extended address of 16 bytes. But with the current code some parts of the domain where stored. For example on my machine something like "010203040506.loc" was saved when the full name was ""010203040506.local.".